1-building-an-instagram-hashtag-typeahead-in-javascript/post.md

```diff
@@ -0,0 +1,202 @@
+*See it in action at [tailwindapp.com](http://tailwindapp.com)*
+
+Sounds pretty straightforward, right? Turns out, there are several technical challenges and constraints that we had to consider.
+
+Let’s walk through the main challenges of building and rendering a typeahead.
+
+## How to detect whether a hashtag is being created/edited
+
+When a user is in the process of editing an Instagram post caption, one of several different things can be true:
+
+1. The user might be typing a new hashtag or editing an existing one
+
+1. The user‘s caret might be focused on a hashtag, but the user might not actually be editing the hashtag at the moment
+
+1. The user could be editing the post caption but not typing a hashtag
+
+In which cases would we want to display a typeahead?
+
+As it turns out, we would like to display the typeahead for case #1, but not cases #2 and #3. Typeaheads are powerful components, but they should be used only in cases where you *know* they are helpful to your user, which is why we only want to show our Typeahead when the user is typing a hashtag.
+
+## To render the typeahead or not to render — That is the question
+
+In order to determine whether the typeahead should be shown, we need to get the hashtag (we’ll call that the `activeHashtag`) that the user is creating/editing. To do that, we need:
+
+1. A `keydown` event handler that will fire every time the user presses a key (or a key combination such as `shift + a`)
+
+1. A function called `getActiveHashtag` that takes the following arguments:
+`content` — The content of the post caption that the user is editing. E.g. `Hello #world!`
+
+`key` — The key that the user pressed to trigger this keydown event. We retrieve this from the event object. E.g. if the user presses thea key, then event.key would be a. There’s more information available as part of the event object that can determine whether a key combination was pressed (such as `shift + A`), but in our case, `event.key` is all we care about.
+
+`caretIndex` — In order to determine what part of the post caption the user is editing, we need to know where the caret is. so that we know exactly what part of the post caption the user is editing. E.g. if the user’s caret is here: `#hello worl|` , then the caret index would be `11`. We get this information from `event.target.selectionStart`.
+
+The `getActiveHashtag` function will either return the hashtag the user is editing (e.g. `#worl`), or `null` if no hashtag is actively being edited by the user.
+
+With these arguments in mind and this desired output (either a string that is the hashtag, or `null`), let’s build some pseudocode test cases that we can use to determine whether or not a user is editing the post caption! You should be able to do this with any common testing suite. At Tailwind, we do Test-Driven Development, using [tape](https://www.npmjs.com/package/tape), to reduce bugs and ensure that we’ve planned our implementation correctly before we write the actual code.
```

Comment:
> You should be able to do this with any common testing suite.

seems off to first introduce the tutorial tone here. I would remove as the paragraph seems fine without it.